File: hexameter-core/src/main/java/org/codetome/hexameter/core/internal/impl/layoutstrategy/GridLayoutStrategy.java
package org.codetome.hexameter.core.internal.impl.layoutstrategy;

import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.HexagonalGrid;
import org.codetome.hexameter.core.api.HexagonalGridBuilder;
import rx.Observable;

/**
 * Represents the method of creating a {@link HexagonalGrid} corresponding to a given shape.
 */
public abstract class GridLayoutStrategy {

    /**
     * Fetches a monotonically increasing (from left to right, top to bottom) Set of
     * grid coordinates corresponding to the shape of the requested grid layout.
     *
     * @param builder builder
     *
     * @return observable
     */
    public abstract Observable<CubeCoordinate> fetchGridCoordinates(HexagonalGridBuilder builder);

    /**
     * Checks whether the supplied parameters are valid for the given strategy.
     * <i>For example a hexagonal grid layout only works if the width equals to the height</i>
     *
     * @param gridHeight height
     * @param gridWidth width
     *
     * @return valid?
     */
    public boolean checkParameters(final int gridHeight, final int gridWidth) {
        return gridHeight > 0 && gridWidth > 0;
    }

}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/internal/impl/layoutstrategy/GridLayoutStrategy.kt
package org.hexworks.mixite2.core.internal.impl.layoutstrategy

import org.hexworks.mixite2.core.api.CubeCoordinate
import org.hexworks.mixite2.core.api.HexagonalGridBuilder
import org.hexworks.mixite2.core.api.contract.SatelliteData

/**
 * Represents the method of creating a [org.hexworks.mixite.core.api.HexagonalGrid] corresponding to a given shape.
 */
abstract class GridLayoutStrategy {

    /**
     * Fetches a monotonically increasing (from left to right, top to bottom) Set of
     * grid coordinates corresponding to the shape of the requested grid layout.
     *
     * @param builder builder
     *
     * @return observable
     */
    abstract fun fetchGridCoordinates(builder: HexagonalGridBuilder<out SatelliteData>): Iterable<org.hexworks.mixite2.core.api.CubeCoordinate>

    /**
     * Checks whether the supplied parameters are valid for the given strategy.
     * *For example a hexagonal grid layout only works if the width equals to the height*
     *
     * @param gridHeight height
     * @param gridWidth width
     *
     * @return valid?
     */
    abstract fun checkParameters(gridHeight: Int, gridWidth: Int): Boolean

    protected fun checkCommonCase(gridHeight: Int, gridWidth: Int): Boolean {
        return gridHeight > 0 && gridWidth > 0
    }

    /**
     * Returns the name of this strategy.
     *
     * @return name
     */
    abstract fun getName() : String

}
